### ISSUE TITLE
LDAP Authentication (javax.naming.directory.InitialDirContext.InitialDirContext) with bad password results in 775 LDAP error (lock out) after several minute delay 

### ISSUE THREAD

**Issue by Darghan21** [adoptium/adoptium-support#651]
### Please provide a brief summary of the bug

Java Version: OpenJDK Runtime Environment (Temurin)(build 1.8.0_352-b08)

Windows 10 64-bit

We have a Java Application that allows for LDAP authentication.  We have not made any code changes to this method in years.

We used OracleJDK8, then OpenJ9 Java 8, and recently moved to Temurin 1.8.0_352.

With both OpenJ9 and OracleJDK, there is no delay in processing a bad password and it correctly returns the bad password error from LDAP.

Current (Incorrect) Behavior, resulting in an account lock out on first attempt with LDAP after a couple minutes of waiting:
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.AuthenticationException
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	[LDAP: error code 49 - 80090308: LdapErr: DSID-0C090447, comment: AcceptSecurityContext error, data 775, v3839 ]
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3261)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3207)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2993)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2907)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:347)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(LdapCtxFactory.java:225)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:189)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:243)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:154)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:84)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:695)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.init(InitialContext.java:244)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.<init>(InitialContext.java:216)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)

Correct, Bad Password return from LDAP:
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.AuthenticationException
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	[LDAP: error code 49 - 80090308: LdapErr: DSID-0C090439, comment: AcceptSecurityContext error, data 52e, v4563 ]
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3154)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3100)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2886)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2800)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:319)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:187)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.init(InitialContext.java:244)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.<init>(InitialContext.java:216)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)

### Please provide steps to reproduce where possible

import javax.naming.*;
import javax.naming.directory.*;  
import java.util.Hashtable; 
   
   class LDAPError {
    public static void main( String []args ) {
		DirContext authCtx = null;
		try {
			authCtx = createContext("user","password");
			System.out.print("Completed InitialDirContext call>>>");
		}
        catch(NamingSecurityException nse) {
			System.out.println(nse.getMessage());
        }
        catch(Exception e) {
            System.out.println(e.getMessage());
        }
        finally {
			try {
				if (authCtx != null)
					authCtx.close();
			} catch(Exception e) {
			}
        }
		
		System.out.print("All Done!!");
		
    }
	
	   private static DirContext createContext(String user, String password) throws NamingException {
        // Set up the environment for creating the initial context
        Hashtable env = new Hashtable();
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
        env.put(Context.PROVIDER_URL, "put_ldap_url_here");

        // Authenticate using simple
        env.put(Context.SECURITY_AUTHENTICATION, "simple");
        env.put(Context.SECURITY_PRINCIPAL, "username");
        env.put(Context.SECURITY_CREDENTIALS, "badpassword");

        // This line authenticates
        System.out.print("Starting InitialDirContext call>>>");
        return new InitialDirContext(env); //This is the line that delays and then returns the exception for lock out
    }
}

### Expected Results

We expect to receive the below error code indicating bad password from LDAP with a quick response within seconds, not minutes:

Correct, Bad Password return from LDAP:
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.AuthenticationException
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	[LDAP: error code 49 - 80090308: LdapErr: DSID-0C090439, comment: AcceptSecurityContext error, data 52e, v4563 ]
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3154)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3100)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2886)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2800)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:319)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:187)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.init(InitialContext.java:244)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.InitialContext.<init>(InitialContext.java:216)
ERROR	2022-12-12 12:32:31,586	SYSTEM	stack-trace	javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)

### Actual Results

After a couple minutes of delay, the LDAP password is locked out and the following exception is returned:

ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.AuthenticationException
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	[LDAP: error code 49 - 80090308: LdapErr: DSID-0C090447, comment: AcceptSecurityContext error, data 775, v3839 ]
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3261)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3207)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2993)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2907)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:347)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(LdapCtxFactory.java:225)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:189)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:243)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:154)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:84)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:695)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.init(InitialContext.java:244)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.InitialContext.<init>(InitialContext.java:216)
ERROR	2022-12-12 12:54:23,934	SYSTEM	stack-trace	javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)

### What Java Version are you using?

openjdk version "1.8.0_352" OpenJDK Runtime Environment (Temurin)(build 1.8.0_352-b08) OpenJDK 64-Bit Server VM (Temurin)(build 25.352-b08, mixed mode)

### What is your operating system and platform?

Windows 10 x64

### How did you install Java?

Download from Adoptium website

### Did it work before?

```shell
This worked in the OracleJRE8:
java version "1.8.0_162"
Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.162-b12, mixed mode)

and in Eclipse OpenJ9:
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (build 1.8.0_222-b10)
Eclipse OpenJ9 VM (build openj9-0.15.1, JRE 1.8.0 Windows 10 amd64-64-Bit Compressed References 20190717_421 (JIT enabled, AOT enabled)
OpenJ9   - 0f66c6431
OMR      - ec782f26
JCL      - f147086df1e based on jdk8u222-b10)
```

### Did you test with other Java versions?

```shell
I did not test other Temurin Versions
```

**Comment by Darghan21** [comment]
Upon some further debugging, we suspect the for loop searching endpoints may be the culprit (this code is markedly different from the other versions we tested):

LdapCtxFactory:

private static DirContext getUsingURL(String url, Hashtable<?,?> env)
            throws NamingException
    {
        try {
            LdapDnsProviderResult r =
                LdapDnsProviderService.getInstance().lookupEndpoints(url, env);
            LdapCtx ctx;
            NamingException lastException = null;

            /*
             * Prior to this change we had been assuming that the url.getDN()
             * should be converted to a domain name via
             * ServiceLocator.mapDnToDomainName(url.getDN())
             *
             * However this is incorrect as we can't assume that the supplied
             * url.getDN() is the same as the dns domain for the directory
             * server.
             *
             * This means that we depend on the dnsProvider to return both
             * the list of urls of individual hosts from which we attempt to
             * create an LdapCtx from *AND* the domain name that they serve
             *
             * In order to do this the dnsProvider must return an
             * {@link LdapDnsProviderResult}.
             *
             */
            for (String u : r.getEndpoints()) {
                try {
                    ctx = getLdapCtxFromUrl(
                            r.getDomainName(), url, new LdapURL(u), env);
                    return ctx;
                } catch (NamingException e) {
                    // try the next element
                    lastException = e;
                }
            }

            if (lastException != null) {
                throw lastException;
            }

            // lookupEndpoints returned an LdapDnsProviderResult with an empty
            // list of endpoints
            throw new NamingException("Could not resolve a valid ldap host");
        } catch (NamingException e) {
            // lookupEndpoints(url, env) may throw a NamingException, which
            // there is no need to wrap.
            throw e;
        } catch (Exception e) {
            NamingException ex = new NamingException();
            ex.setRootCause(e);
            throw ex;
        }
    }

**Comment by Darghan21** [comment]
This may or may not actually be a bug upon some further testing.

We had an LDAP configuration that did not specify a server, so the loop looking at endpoints kept running despite the bad password and locked out the account with multiple attempts.  Adding a specific server in to the LDAP url caused the failure as expected without a log out, because it attempted only one login.  It does seem like it would make sense to fail with a bad password/bad login on the first attempt?

**Comment by jerboaa** [comment]
Seems related to https://bugs.openjdk.org/browse/JDK-8275535, which isn't in OpenJDK 8u.

**Comment by jerboaa** [comment]
This should be fixed in OpenJDK `17.0.6`, `11.0.18` and `8u372` (which is April 2023). Feel free to test an EA build.
